include/eeprom/core/write_protect.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+namespace eeprom {
+namespace write_protect {
+
+/**
+ * Interface to write protect pin. Must be implemented in FW and Simulation
+ */
+class WriteProtectPin {
+  public:
+    WriteProtectPin() = default;
+    WriteProtectPin(const WriteProtectPin&) = default;
+    WriteProtectPin(WriteProtectPin&&) = default;
+    auto operator=(WriteProtectPin&&) -> WriteProtectPin& = default;
+    auto operator=(const WriteProtectPin&) -> WriteProtectPin& = default;
+    virtual ~WriteProtectPin() = default;
+
+    /**
+     * Change the write protect status.
+     * @param enabled true to inhibit writing
+     */
+    virtual void set(bool) = 0;
+};
+
+/**
+ * Class to manages the state of the write protect pin across multiple
+ * transactions.
+ *
+ * Consecutive write requests will want to have write protection disabled. The
+ * enable should happen when all the write transactions have completed.
+ *
+ * This class keeps the eeprom from having to keep track of the write protection
+ * state.
+ */
+class WriteProtector {
+  public:
+    /**
+     * Constructor
+     * @param pin
+     */
+    explicit WriteProtector(WriteProtectPin& pin) : pin{pin} {}
+
+    /**
+     * Disable the write protect
+     */
+    void disable() {
+        if (count++ == 0) {
```

Comment:
Could you explain the logic of the disable count?

Reply:
The eeprom needs to have write protection disabled for all writes. I want to disable write protect at the start of a write and re-enable it when there are no more writes. 

Due to the asynchronous nature of the i2c task it can be tricky to know when this happens. There could be multiple write requests in the queue simultanously. 

The eeprom task calls disable whenever it gets a write to eeprom request. It then calls enable whenever the write request completes. 

So image two write requests arrive:
eeprom task processes first request -> disables write protect (changes pin state, disable count is 1)
eeprom task processes second request -> disables write protect (does nothing to pin, only increments disable count to 2)
first request completes -> enables write protect (does nothing to pin, decrement count to 1)
second request completes -> enables write protect (changes pin state, decrement count to 0)